File: src/01_data_structures/06_trees/03_trie/implementations/ts/trie.test.ts
import { Trie } from "./trie";

describe("Trie()", () => {
  test("Trie is instantiated", async () => {
    const trie = new Trie();

    expect(trie).toEqual({ root: { end: false, children: {} } });
  });
});

describe("Insert()", () => {
  test("Inserts data into a trie", async () => {
    const trie = new Trie().insert("ab");

    expect(trie).toEqual({
      root: {
        end: false,
        children: {
          a: {
            end: false,
            children: {
              b: {
                end: true,
                children: {},
              },
            },
          },
        },
      },
    });
  });

  test("Inserts empty string into a trie", async () => {
    const trie = new Trie().insert("");
    expect(trie).toEqual({
      root: {
        end: true,
        children: {},
      },
    });

    trie.insert("a");
    expect(trie).toEqual({
      root: {
        end: true,
        children: {
          a: {
            end: true,
            children: {},
          },
        },
      },
    });

    trie.insert("b");
    expect(trie).toEqual({
      root: {
        end: true,
        children: {
          a: {
            end: true,
            children: {},
          },
          b: {
            end: true,
            children: {},
          },
        },
      },
    });

    trie.insert("");
    expect(trie).toEqual({
      root: {
        end: true,
        children: {
          a: {
            end: true,
            children: {},
          },
          b: {
            end: true,
            children: {},
          },
        },
      },
    });
  });
});

describe("Search()", () => {
  test("Searches a word from an empty trie", async () => {
    const trie = new Trie();
    const found = trie.search("a");

    expect(found).toBe(false);
  });

  test("Searches a single character word from a populated trie", async () => {
    const trie = new Trie().insert("a");
    const found = trie.search("a");

    expect(found).toBe(true);
  });

  test("Searches a word from a populated trie", async () => {
    const trie = new Trie().insert("ab");
    const found = trie.search("ab");

    expect(found).toBe(true);
  });

  test("Searches a word from a populated trie with several words", async () => {
    const trie = new Trie().insert("ab").insert("abc").insert("ac");

    expect(trie.search("ab")).toBe(true);
    expect(trie.search("abc")).toBe(true);
    expect(trie.search("ac")).toBe(true);
  });
});

describe("Delete()", () => {
  test("Deletes a word from an empty trie", async () => {
    const trie = new Trie().delete("abc");

    expect(trie).toEqual({
      root: {
        end: false,
        children: {},
      },
    });
  });

  test("Deletes a word from a populated trie", async () => {
    const trie = new Trie();

    expect(trie).toEqual({
      root: {
        end: false,
        children: {},
      },
    });

    trie.insert("abc");
    expect(trie).toEqual({
      root: {
        end: false,
        children: {
          a: {
            end: false,
            children: {
              b: {
                end: false,
                children: {
                  c: {
                    end: true,
                    children: {},
                  },
                },
              },
            },
          },
        },
      },
    });

    trie.delete("abc");
    expect(trie).toEqual({
      root: {
        end: false,
        children: {},
      },
    });
  });

  test("Deletes a word from a trie with a longer word", async () => {
    const trie = new Trie().insert("abc").insert("abcd");

    expect(trie).toEqual({
      root: {
        end: false,
        children: {
          a: {
            end: false,
            children: {
              b: {
                end: false,
                children: {
                  c: {
                    end: true,
                    children: {
                      d: {
                        end: true,
                        children: {},
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    });

    trie.delete("abc");
    expect(trie).toEqual({
      root: {
        end: false,
        children: {
          a: {
            end: false,
            children: {
              b: {
                end: false,
                children: {
                  c: {
                    end: false,
                    children: {
                      d: {
                        end: true,
                        children: {},
                      },
                    },
                  },
                },
              },
            },
          },
        },
      },
    });

    trie.delete("abcd");
    expect(trie).toEqual({
      root: {
        end: false,
        children: {},
      },
    });
  });
});

describe("Autocomplete()", () => {
  test("...", async () => {
    const trie = new Trie()
      .insert("abcd")
      .insert("abce")
      .insert("abcf")
      .insert("ayxz");

    const autocomplete = trie.suggest("ab");

    expect(autocomplete).toEqual(["abcd", "abce", "abcf"]);
  });
});

File: src/01_data_structures/06_tree/03_trie/implementations/go/trie_test.go
package trie

import (
	"reflect"
	"testing"
)

func TestTrieIsInstantiated(t *testing.T) {
	trie := NewTrie()

	expected := &Trie{
		Root: NewNode(),
	}

	if !reflect.DeepEqual(trie, expected) {
		t.Errorf("Expected %v, got %v", expected, trie)
	}
}

func TestInsertsDataIntoTrie(t *testing.T) {
	trie := NewTrie().Insert("ab")

	expected := &Trie{
		Root: &Node{
			End: false,
			Children: map[string]*Node{
				"a": {
					End: false,
					Children: map[string]*Node{
						"b": {
							End:      true,
							Children: map[string]*Node{},
						},
					},
				},
			},
		},
	}

	if !reflect.DeepEqual(trie, expected) {
		t.Errorf("Expected %v, got %v", expected, trie)
	}
}

func TestInsertsEmptyStringIntoTrie(t *testing.T) {
	trie := NewTrie().Insert("")

	expected1 := &Trie{
		Root: &Node{
			End:      true,
			Children: map[string]*Node{},
		},
	}
	if !reflect.DeepEqual(trie, expected1) {
		t.Errorf("Expected %v, got %v", expected1, trie)
	}

	trie.Insert("a")
	trie.Insert("b")

	expected2 := &Trie{
		Root: &Node{
			End: true,
			Children: map[string]*Node{
				"a": {
					End:      true,
					Children: map[string]*Node{},
				},
				"b": {
					End:      true,
					Children: map[string]*Node{},
				},
			},
		},
	}
	if !reflect.DeepEqual(trie, expected2) {
		t.Errorf("Expected %v, got %v", expected2, trie)
	}

	trie.Insert("")
	if !reflect.DeepEqual(trie, expected2) {
		t.Errorf("Expected %v, got %v", expected2, trie)
	}
}

func TestSearchEmptyTrie(t *testing.T) {
	trie := NewTrie()
	found := trie.Search("a")

	if !reflect.DeepEqual(found, false) {
		t.Errorf("Expected %v, got %v", found, false)
	}
}

func TestSearchSingleCharacter(t *testing.T) {
	trie := NewTrie()
	trie.Insert("a")
	found := trie.Search("a")

	if !reflect.DeepEqual(found, true) {
		t.Errorf("Expected %v, got %v", found, true)
	}
}

func TestSearchMultipleCharacter(t *testing.T) {
	trie := NewTrie()
	trie.Insert("ab")
	found := trie.Search("ab")

	if !reflect.DeepEqual(found, true) {
		t.Errorf("Expected %v, got %v", found, true)
	}
}

func TestSearchMultipleCharacterSeveralWords(t *testing.T) {
	trie := NewTrie()
	trie.Insert("ab").Insert("ab").Insert("abc").Insert("ac")

	foundAb := trie.Search("ab")
	if !reflect.DeepEqual(foundAb, true) {
		t.Errorf("Expected %v, got %v", foundAb, true)
	}

	foundAbc := trie.Search("abc")
	if !reflect.DeepEqual(foundAbc, true) {
		t.Errorf("Expected %v, got %v", foundAbc, true)
	}

	foundAc := trie.Search("ac")
	if !reflect.DeepEqual(foundAc, true) {
		t.Errorf("Expected %v, got %v", foundAc, true)
	}
}

func TestDeleteFromEmptyTrie(t *testing.T) {
	trie := NewTrie().Delete("abc")

	expected := &Trie{
		Root: &Node{
			End:      false,
			Children: map[string]*Node{},
		},
	}

	if !reflect.DeepEqual(trie, expected) {
		t.Errorf("Expected %+v, got %+v", expected, trie)
	}

}

func TestDeleteFromPopulatedTrie(t *testing.T) {
	trie := NewTrie()

	// Initial state
	expectedEmpty := &Trie{
		Root: &Node{
			End:      false,
			Children: map[string]*Node{},
		},
	}
	if !reflect.DeepEqual(trie, expectedEmpty) {
		t.Errorf("Expected %+v, got %+v", expectedEmpty, trie)
	}

	trie.Insert("abc")
	expectedWithABC := &Trie{
		Root: &Node{
			End: false,
			Children: map[string]*Node{
				"a": {
					End: false,
					Children: map[string]*Node{
						"b": {
							End: false,
							Children: map[string]*Node{
								"c": {
									End:      true,
									Children: map[string]*Node{},
								},
							},
						},
					},
				},
			},
		},
	}
	if !reflect.DeepEqual(trie, expectedWithABC) {
		t.Errorf("Expected %+v, got %+v", expectedWithABC, trie)
	}

	trie.Delete("abc")

	if !reflect.DeepEqual(trie, expectedEmpty) {
		t.Errorf("Expected %+v, got %+v", expectedEmpty, trie)
	}
}

func TestDeleteFromTrieWithLongerWord(t *testing.T) {
	trie := NewTrie()
	trie.Insert("abc").Insert("abcd")

	expectedBeforeDeletes := &Trie{
		Root: &Node{
			End: false,
			Children: map[string]*Node{
				"a": {
					End: false,
					Children: map[string]*Node{
						"b": {
							End: false,
							Children: map[string]*Node{
								"c": {
									End: true,
									Children: map[string]*Node{
										"d": {
											End:      true,
											Children: map[string]*Node{},
										},
									},
								},
							},
						},
					},
				},
			},
		},
	}
	if !reflect.DeepEqual(trie, expectedBeforeDeletes) {
		t.Errorf("Expected %+v, got %+v", expectedBeforeDeletes, trie)
	}

	trie.Delete("abc")
	expectedAfterFirstDelete := &Trie{
		Root: &Node{
			End: false,
			Children: map[string]*Node{
				"a": {
					End: false,
					Children: map[string]*Node{
						"b": {
							End: false,
							Children: map[string]*Node{
								"c": {
									End: false,
									Children: map[string]*Node{
										"d": {
											End:      true,
											Children: map[string]*Node{},
										},
									},
								},
							},
						},
					},
				},
			},
		},
	}
	if !reflect.DeepEqual(trie, expectedAfterFirstDelete) {
		t.Errorf("Expected %+v, got %+v", expectedAfterFirstDelete, trie)
	}

	trie.Delete("abcd")
	expectedEmpty := &Trie{
		Root: &Node{
			End:      false,
			Children: map[string]*Node{},
		},
	}
	if !reflect.DeepEqual(trie, expectedEmpty) {
		t.Errorf("Expected %+v, got %+v", expectedEmpty, trie)
	}
}

func TestSuggestsReturnsValidOptionsFromPrefix(t *testing.T) {
	trie := NewTrie().Insert("abcd").Insert("abce").Insert("abcf").Insert("ayxz")
	autocomplete := trie.Suggest("ab")
	expected := []string{"abcd", "abce", "abcf"}

	if !reflect.DeepEqual(autocomplete, expected) {
		t.Errorf("Expected %v, got %v", expected, autocomplete)
	}
}
